src/components/structures/ScrollPanel.js

```diff
@@ -131,8 +148,53 @@ module.exports = React.createClass({
         if (sn.scrollTop < sn.clientHeight) {
             // there's less than a screenful of messages left - try to get some
             // more messages.
-            this.props.onFillRequest(true);
+            this._maybeFill(true);
         }
+        if (sn.scrollTop > sn.scrollHeight - sn.clientHeight * 2) {
+            this._maybeFill(false);
+        }
+    },
+
+    // check if there is already a pending fill request. If not, set one off.
+    _maybeFill: function(backwards) {
+        var dir = backwards ? 'b' : 'f';
+        if (this._pendingFillRequests[dir]) {
+            if (DEBUG_SCROLL) {
+                console.log("ScrollPanel: Already a "+dir+" fill in progress - not starting another");
+            }
+            return;
+        }
+
+        if (DEBUG_SCROLL) {
+            console.log("ScrollPanel: starting "+dir+" fill");
+        }
+
+        // onFillRequest can end up calling us recursively (via onScroll
+        // events) so make sure we set this before firing off the call. That
+        // does present the risk that we might not ever actually fire off the
+        // fill request, so wrap it in a try/catch.
+        this._pendingFillRequests[dir] = true;
+        var r;
+        try {
+             r = this.props.onFillRequest(backwards);
```

Comment:
I don't like `r` as a variable name, why is it even called that? For "request" maybe? I would intuitively call it `fillPromise` but this is complicated because you allow multiple return types, which I've already commented on disliking (this is one reason why; it makes "what does the return type represent" really hard to name).